### PR TITLE
Add additional track sprites to giga coaster

### PR DESCRIPTION
**Pull request by X123M3-256**
This isn't complete and not ready for merge, it's maybe 98% done but I'm probably not going to finish it anytime soon so I am submitting a PR so that it is available in case anyone else wants to try and finish it. The following tasks remain to be done:

* Adjusting supports where they fail to connect with the track (mostly done for the giga coaster, not started for the mini coaster)

* Fixing the small loop sprites and bounding boxes (they currently glitch badly, although removing the small loops entirely is also an option)

* Removing the track pieces which won't work with existing vehicle sprites  (which is most of them) from the available track pieces and make them available with cheats only.

* I had planned to upgrade the vehicle paint code to improve the selection of fallback sprites for the diagonal banks - if it used the sprites for banked sloped turns, rather than the unbanked diagonal sprites, it might look acceptable enough for the diagonal sloped turns to be available by default. This wouldn't be necessary for merge, however, as the diagonal pieces can just be left disabled.

I would appreciate if someone could test to ensure that the branch is functional and I didn't forget to commit any sprites.

### CODE REVIEW

**Comment by karst** [comment]
I am testing using the mega coaster sprites of yours to look for bounding box issues. But I haven't found much major. There's a minor issue with the medium loop left.

image

image

The supports for the mini coaster i'll gladly help out with. I haven't noticed anything for the giga.

> Removing the track pieces which won't work with existing vehicle sprites (which is most of them) from the available track pieces and make them available with cheats only.

This won't need to be exact as pieces won't be available anyhow when a ride vehicle doesn't support the right groups. Though we definitely need to disable most that don't make sense for giga. Then we can enable them in the future for the "Mega Coaster", "Multi-Launch coaster" or "Multi-Looping coaster" or whatever the idea was. Once those get added it'd probably be a good time to disable LSMs (And perhaps even chainlifts as both Milli and I305 use cablelift) on the Giga.

**Comment by karst** [comment]
image
image
it seems the small loop is missing some sprites for both

**Comment by kyphii** [comment]
Noticed that vertical track is inaccessible unless built off the Quarter Loop or off ride type change. Is this intentional?

**Comment by X123M3-256** [comment]
> image image it seems the small loop is missing some sprites for both

It's not missing sprites, they are getting clipped because the sprites extend outside the borders of their tile element. This is normally a sprite mask issue, but the system I use to do the automatic splitting isn't up to doing this correctly. They will need to be fixed manually or else removed entirely.

The bounding boxes also appear broken, and it's missing a support because I copied the code from the LIM coaster that is apparently also missing that support which is present on all other rides.

**Comment by X123M3-256** [comment]
> Noticed that vertical track is inaccessible unless built off the Quarter Loop or off ride type change. Is this intentional?

This was not intentional, I forgot to enable them. However, all the track pieces that are currently enabled should not be, they should be moved to the ExtraTrackPieces so they only show up when the cheat is enabled.

The exception would be large steep to flat on the mini coaster, and the banked turn to gentle slope piece on both - those can be added by default. As I mentioned, with some changes to the vehicle paint code, the large sloped turns potentially could also be added by default.

**Comment by 733737** [comment]
can't we simply add one element at a time? So split this pr into 20 separate prs, and that way we can add elements that need less fixes sooner, and save elements that need more work for later?

**Comment by karst** [comment]
> can't we simply add one element at a time? So split this pr into 20 separate prs, and that way we can add elements that need less fixes sooner, and save elements that need more work for later?

I don't think getting the support right will cost me that much work. A few weeks max. So i doubt it'll be necessary.

**Comment by Realsteel89** [comment]
good work just takes time

**Comment by Legotown55** [comment]
just tested the new banked curves on the giga coaster and they work fine

**Comment by karst** [comment]
Note to self not to forget:

* [ ] Park version increase
* [ ] Network version increase
* [ ] Changelog entry
* [ ] Import functions keep track pieces invisible upon importing older save files than this version.

**Comment by karst** [comment]
Should I split this PR into two? One for the giga one for the mini?

**Comment by Legotown55** [comment]
Will the loops come back on the mini coaster

**Comment by karst** [comment]
yes i will create a new pr for it in a bit

**Comment by Realsteel89** [comment]
Boath he want to fix I work on the maurer too for it only need to fix the pngs a bit the loop self works

**Comment by Legotown55** [comment]
What is going on with this PR

**Comment by Aeromachinator** [comment]
> What is going on with this PR

Self-paced progress is what's going on. This project is not something that people work on full-time, but rather when they get the chance to. That's why things move around relatively slowly around here, where patience is the best course of action.

**Comment by karst** [comment]
This PR is ready for review. I imagine I have forgotten something so just let me know and i'll try working out fixes.

**Comment by Legotown55** [comment]
Why is this rivew takeing so long

**Comment by ZehMatt** [comment]
> Why is this rivew takeing so long

I believe you have been warned in the past to keep comments like that to yourself especially when you are not contributing in any meaningful way, let this be your last warning.

**Comment by Realsteel89** [comment]
@Legotown55 it just takes him time to post-process the images individually from the small zero g roll and the vertical loops.
I know myself how crappy that is because I did it with my track too. to fix all tooks about 8h for me.
I will help him

**Comment by Broxzier** [comment]
It's pointless to respond to legotown. They never reply, all they do is complain about random PRs not being merged yet.

**Comment by X123M3-256** [comment]
> Why is this rivew takeing so long

Because it's not just waiting for review, it was never completed. I gave up working on it and submitted unfinished in case anyone else wanted to finish it. It looks like @karst has made some progress with it but it still isn't ready to merge.

**Comment by karst** [comment]
https://cdn.discordapp.com/attachments/264137540670324737/1118256169807511632/Giga_Sprite_Test.park here's a test park with some coasters

**Comment by ZehMatt** [comment]
image

I don't think this is quite ready yet.

**Comment by karst** [comment]
> image
> 
> I don't think this is quite ready yet.

Yea noted this! Working on it as we speak. I got it fixed in looks now the bounding boxes left to do on that.

**Comment by Realsteel89** [comment]
The last screens looked as an outdated version on it or did you reset it @karst

**Comment by Legotown55** [comment]
Vagabond Falls 2023-06-17 11-33-24
Can you deal with this bug in the screenshot

**Comment by karst** [comment]
> Vagabond Falls 2023-06-17 11-33-24 Can you deal with this bug in the screenshot

This isn't related to this PR.

**Comment by Realsteel89** [comment]
> screenshot

What bug it looked that you don't choose an giga track car into your list that have no ride preview picture. That's why you had no preview image to giga track there.

File: src/openrct2/sprites.h
Comment (X123M3-256) [inline src/openrct2/sprites.h:1219]: Wait what did you do to the loop sprites? There's not supposed to be additional sprites added, there should just be 32.

Comment (karst) [reply src/openrct2/sprites.h:1219]: Added a few extra, cutout, sprites to fill on the spaces that were outside the loops drawing boundry. They now show in the correct spot, but bounding boxes are wrong.

Comment (X123M3-256) [reply src/openrct2/sprites.h:1219]: Yeah, but adding extra sprites isn't the best way to do this ... I mean, it can work, but I'd rather keep the sprite count and paint code the same as the vanilla rides because that would make a future refactor simpler ... and also, adding new sprites means more bounding boxes to tweak.

The way to fix the clipping is to move the parts of the sprites that clip to other sprites that won't, to match the way the vanilla loop sprites are split. I couldn't figure out how to get my renderer to output the correct splitting automatically, like it usually would - this isn't the first track piece that had to be split manually like this I just normally wouldn't submit a PR with such an issue. I'm going to try and fix this properly now. If I'm correct, this is the only thing left for the giga track?

**Comment by Legotown55** [comment]
Why has this PR nont been reviewed yet?

**Comment by X123M3-256** [comment]
> Why has this PR nont been reviewed yet?

Because it isn't ready for review.  I pretty much abandoned it months ago and I decided to submit what I had so that what I did is available for anyone that might want to do something with it ... like I wrote in the description.

It would probably only take few days to finish, it is really just the support heights that need adjusting now - but I'm not working on it and I don't know if I will come back to it.

**Comment by Legotown55** [comment]
Tanks for the reply

**Comment by Legotown55** [comment]
X123M3-256 Please compleat this PR thanks

**Comment by X123M3-256** [comment]
I fixed all remaining supports and the glitches on the medium half loop and the large sloped turns. I think this is ready for review now.

**Review by Legotown55 — APPROVED** [review]
Thi PR is looking good for merging

**Review by ZehMatt — COMMENTED** [review]
See comment, will try to playtest a bit before I approve it, its quite a lot of code so I'll have to see if I spot anything odd in-game.

**Comment by Legotown55** [comment]
Can the devs get this PR murged very soon

**Comment by ZehMatt** [comment]
I did a bit of testing and found a few things that definitely feel a bit odd:

https://github.com/OpenRCT2/OpenRCT2/assets/5415177/7bac1298-e79d-4a8b-bc63-92bb283f01b4

https://github.com/OpenRCT2/OpenRCT2/assets/5415177/730481ee-e29e-425a-bcfb-1b3ffbb47efb

Here the track is sometimes drawn over the vehicle

https://github.com/OpenRCT2/OpenRCT2/assets/5415177/d3677b04-c047-4205-b709-81303f87ada2

Here the wheels seem to be quite disconnected from the track

image

Here the vehicle is drawn over the track

**Comment by Realsteel89** [comment]
grafik
I think its no problem with the track more with the trains the cars are not centered maybe if he render it

**Comment by X123M3-256** [comment]
@ZehMatt 

> Here the track is sometimes drawn over the vehicle

That's an existing vanilla track sprite. The glitching is a bug and I can fix it if you want but it's not related to the code I added, it would mean changing the existing code. It's one of those things where I really wish there was less code duplication because the bounding boxes are generally the same between ride types and there's probably a lot of rides with this issue.

> Here the wheels seem to be quite disconnected from the track

There's a few possible causes:

* The nominal position of the car is in the center, so in a tight roll, the car's' bank angle will be determined by that center point and won't match the wheels, where the track is banked more. This issue can also cause the wheels to lift off the track on a hill crest. A fix would require changes to the vehicle paint code, and for the vehicles themselves to be modified with an offset for the position of the wheels. It's definitely outside the scope of this PR but it might be a nice addition.

* The sprites for the barrel rolls represent a simple rotation about the cars longitudinal axis. This does not match the track - the track in a barrel roll is at an angle to the centerline of the roll - the car can't actually rotate about its longitudinal axis except on pipeline coasters. This is particularly visible in game as the barrel rolls are tighter than they would be IRL. It would be possible to add a new sprite group with more accurate angles - however, with the current system, that same set of sprites works for inline twists, barrel rolls, and could also work for longer barrel rolls or rolls starting from banked if those were added later. If you add sprites to match the exact angle of the standard barrel rolls, they are only right for those rolls. For this reason, I designed the zero G sprites similarly - they don't perfectly match the zero G rolls I made, but are equally useful for any size zero G, zero G stalls, inclined barrel rolls, and other such pieces that might be added in future.

* It's also possible that the barrel roll subposition data is wrong/doesn't match the sprite. When desigining the shape of the track pieces, I made them to match the vanilla sprites (which they do almost perfectly) but I never checked to see how closely the subposition data matches the sprite. It's possible that it doesn't, in which case genererating new subposition data would be straightforward but it has the potential to break existing designs.

> Here the vehicle is drawn over the track

That one definitely needs fixing.

**Comment by ZehMatt** [comment]
> That's an existing vanilla track sprite. The glitching is a bug and I can fix it if you want but it's not related to the code I added, it would mean changing the existing code. It's one of those things where I really wish there was less code duplication because the bounding boxes are generally the same between ride types and there's probably a lot of rides with this issue.

Wasn't quite sure whats new and whats old, just some observation. I also agree with the duplicate code, ideally each track piece would be an object so we don't have to fiddle around with the code and just generalize the drawing with parameters from the objects.

> The sprites for the barrel rolls represent a simple rotation about the cars longitudinal axis. This does not match the track - the track in a barrel roll is at an angle to the centerline of the roll - the car can't actually rotate about its longitudinal axis except on pipeline coasters. This is particularly visible in game as the barrel rolls are tighter than they would be IRL. It would be possible to add a new sprite group with more accurate angles - however, with the current system, that same set of sprites works for inline twists, barrel rolls, and could also work for longer barrel rolls or rolls starting from banked if those were added later. If you add sprites to match the exact angle of the standard barrel rolls, they are only right for those rolls. For this reason, I designed the zero G sprites similarly - they don't perfectly match the zero G rolls I made, but are equally useful for any size zero G, zero G stalls, inclined barrel rolls, and other such pieces that might be added in future.

I somewhat suspected that this is a limitation, perhaps we can just ever so slightly adjust it a bit so it won't look as floaty until we figure out a proper way.

The rest seems mostly fine, also please address the comment I left.

**Comment by Legotown55** [comment]
Why has this workflow gone stail

File: src/openrct2/ride/Track.cpp
Comment (X123M3-256) [reply src/openrct2/ride/Track.cpp]: Should be fixed

**Comment by X123M3-256** [comment]
I have fixed the glitching on vertical twists. I am going to assume unless told otherwise stated that glitches in the track pieces I didn't add will not be fixed in this PR.

Since the vertical twist bounding boxes would have been copied originally from the twister coaster, it might be worth investigating whether the glitch also occurs there, and if it doesn't why not - I don't like having the bounding boxes different for different ride types without a good reason, but at this point there have been enough such changes that is is probably easiest to deal with this when the track paint code is finally refactored.

**Comment by duncanspumpkin** [comment]
> I have fixed the glitching on vertical twists. I am going to assume unless told otherwise stated that glitches in the track pieces I didn't add will not be fixed in this PR.
> 
> Since the vertical twist bounding boxes would have been copied originally from the twister coaster, it might be worth investigating whether the glitch also occurs there, and if it doesn't why not - I don't like having the bounding boxes different for different ride types without a good reason, but at this point there have been enough such changes that is is probably easiest to deal with this when the track paint code is finally refactored.

Is this PR ready then for final reviews?

**Review by karst — APPROVED** [review]
I already playtested this PR quite intensively beforehand. All the glitches through the tracks that have been mentioned and found have definitely been fixed. Looks buttersmooth, incredible work!

**Review by duncanspumpkin — APPROVED** [review]

**Pull request merged**
